### PR TITLE
feat(security): Add fsGroup to all deployments

### PR DESCRIPTION
**Pull request by karlskewes**
Building on #25 

https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-pod

This is required when Kubernetes mounts in tokens like the AWS IRSA token into
containers, otherwise they get mounted owned by `root` and `clouddriver` can't
read them.
Similar story for other volume mounting though.

> Since `fsGroup` field is specified, all processes of the container are also
part of the supplementary group ID 2000. The owner for volume `/data/demo` and
any files created in that volume will be Group ID 2000.

Assists with: https://github.com/spinnaker/spinnaker/issues/3760

We use `fsGroup` but with ID same as the current Spinnaker service Dockerfile.

### CODE REVIEW

**Comment by karlskewes** [comment]
wrt: IRSA - see end of this comment for required pod spec (what we use as well for clouddriver): https://github.com/spinnaker/spinnaker/issues/5521#issue-573376779

**Comment by maggieneterval** [comment]
Thanks for this PR, @kskewes! 🎉 💟 

I've been trying to learn more about best practices when it comes to configuring a non-root user in Dockerfiles and pod `securityContext` in Deployments. Here's my understanding so far:

- Docker's docs on [Dockerfile best practices](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/) docs say to "Consider an explicit UID/GID" since "Users and groups in an image are assigned a non-deterministic UID/GID in that the “next” UID/GID is assigned regardless of image rebuilds." Spinnaker's Dockerfiles currently do not assign an explicit UID, and rely on the fact that the "next" UID will always be 1000 as that is the first non-root UID to be assigned in the base images we build from. Should we consider explicitly assigning the spinnaker user to UID 1000 in each Spinnaker microservice Dockerfile, or is it pretty standard to rely on that assumption and then couple the UID specified in the Deployment to that either specified or assumed in the Dockerfile?
- Kubernetes' docs on [Deployment best practices](https://kubernetes.io/blog/2016/08/security-best-practices-kubernetes-deployment/) recommend setting the `securityContext` fields that @micnncim set in the [PR](https://github.com/spinnaker/kustomization-base/pull/25) on which this builds. That said, I'm now curious if they are strictly necessary if each of Spinnaker's microservice Dockerfiles is already specifying a non-root user; while of course if one were manually running these containers one could override the Dockerfile-specified user to run as root, is that still a vulnerability when running the container as part of a Kubernetes Deployment?

Regardless of the answers to those questions, I agree we need to set `fsGroup` to the spinnaker UID, so we should merge this, but for my own understanding definitely interested in thoughts on the necessity of `runAsUser` and `runAsNonRoot` when a non-root user is already specified in Dockerfiles. Huge thanks to @kskewes and @micnncim for starting this conversation.

**Review by maggieneterval — APPROVED** [review]

**Comment by karlskewes** [comment]
This is my take on it but I'm sure someone more knowledgeable will be able to explain better and correct me :)

TLDR; Options are:
1. Do everything in `Dockerfile` but get scolded by Kubernetes linters and have questions by OpenShift/users?
2. Do everything in `securityContext` which satisfies linters but have potential mismatch between dev (docker) and deploy (k8s) 
3. Do both << achievable, clear and path of least resistance?

> "Consider an explicit UID/GID"

In [`clouddriver-slim.Dockerfile`](https://github.com/spinnaker/clouddriver/blob/master/Dockerfile.slim#L32) we use the `-S` flag to create a `System` user (beginning from ~100) rather than a regular user (beginning from 1000).
It's possible to specify the user id with `--UID` flag per: https://wiki.alpinelinux.org/wiki/Setting_up_a_new_user#adduser

>  Should we consider explicitly assigning the spinnaker user to UID 1000 in each Spinnaker microservice Dockerfile

This would be optimal because during development you can run the Docker image in the same conditions as might be enforced in Kubernetes via `securityContext` or a `PodSecurityPolicy`.
This becomes particularly important when dealing with files on disk, such as the normal JVM `/tmp` directory usage and also the new Plugin system which downloads and writes files to the filesystem.

https://kubernetes.io/blog/2018/07/18/11-ways-not-to-get-hacked/#8-run-containers-as-a-non-root-user

So ideally not 1000 (1k) but 10000+ (10k+) so there's no chance of colliding with the host OS's users. 

> recommend setting the `securityContext` fields

Setting `runAsNonRoot` does enforce use of a user in the `Dockerfile` and gets to the heart of the matter.
However many linters/security checks will complain if users are not set in the yaml or have low UID, see:
- https://kubesec.io/
- https://github.com/bridgecrewio/checkov/tree/master/tests/kubernetes/checks/example_RootContainersHighUID

**Comment by maggieneterval** [comment]
@kskewes Wow, thanks for another amazing explanation; I always learn so much from your comments.

> Do both << achievable, clear and path of least resistance?

👍 Makes sense.

> So ideally not 1000 (1k) but 10000+ (10k+) so there's no chance of colliding with the host OS's users.

👍 Sounds good. Since the [Spinnaker Platform SIG](https://github.com/spinnaker/governance/blob/master/sig-platform/README.md) owns the Dockerfiles, I'll attend their next sync to get their approval before setting an explicit, 10k+ UID in each. Then I'll update the Kustomizations accordingly. (But I'll merge this PR as-is for now!)

**Comment by karlskewes** [comment]
FWIW I just hit the issue with UID's when I built Fiat locally using the provided Dockerfiles. The new UID was 102 yet the Spinnaker OSS image was 100 so I was unable to install plugins due to file permissions.

**Comment by maggieneterval** [comment]
Thanks for mentioning that, @kskewes. I'm going to get feedback on our proposal to set UIDs in each Dockerfile this Thursday at the Platform SIG. It's often attended by folks working on plugins so I'll share your comment with them as an additional data point in favor.
